### ISSUE TITLE
Array<any> return type showing errors incorrectly

### ISSUE THREAD

**Issue by cyorobert** [facebook/flow#4745]
If this is a bug, please try reproducing using https://flow.org/try/.

^^ I received no errors in the web sandbox, but receive errors using Flow in conjunction with JetBrains as a language setting.

Flow version: 0.53.1
Environment: node.js 7.6.1
IDE: JetBrains PHPStorm (Flow as a language setting)

I have some code that extracts an array of objects with arbitrary data from a database query. It looks like this:

```javascript
//@flow
'use strict'

type BookshelfModel = {
    attributes : any
}
type BookshelfCollection = {
    models : Array<BookshelfModel>
}

export const mapBookshelfCollection = ( thenable : Promise<BookshelfCollection> ) : Promise<Array<any>> => {
    return thenable.then((collection) => collection.models || [])
                   .then((models : Array<BookshelfModel>) => models.map((model) => model.attributes))
                   .catch(() => [])
}
```

My return type in this case is `Promise<Array<any>>` because this method could be used to literally map any collection of models from any database table using the Bookshelf.js ORM.

Using PHPStorm I get the flow error: `Flow: array type. This type is incompatible with object type.`. I'm not sure why I'm getting this error and believe that I'm seeing the error incorrectly.

Any thoughts or suggestions are appreciated. Thank you!!

**Comment by asolove** [comment]
Sorry, this looks like a pretty annoying error. And sadly it seems like Flow is giving the error in the wrong place. This code looks fine by itself, so I would guess it's a problem with one of the places where this function is used. Can you share more about how this is used? Or try removing calls to this function one by one to see which place it's called is generating the error?

**Comment by cyorobert** [comment]
Thanks for the response @asolove. The calling context is always the same:

```javascript
import { mapBookshelfCollection } from './utils'
mapBookshelfCollection(
    SomeBookshelfModel.where({
         some_field : some_value
    })
    .fetchAll()
)
```

The `fetchAll` method returns a `bookshelf.Collection` instance, so I updated my code to include

**Comment by asolove** [comment]
Hmm, I tried this a couple ways and it worked. I think it must be something about how the promise returned from that function is used. But the error message is definitely showing up in the wrong place.

**Comment by cyorobert** [comment]
@asolove Your intuition was spot on! As soon as I fixed my return type in the calling function the error went away. That would have taken quite some time to figure out. I think this specific issue can be closed, but it may be helpful to open another issue about the error being thrown in the wrong place. 

In this case, I believe it's clear that the calling function's return signature should have contained the error rather than the `mapBookshelfCollection`.

**Comment by asolove** [comment]
Yeah, can you share something about how the calling function was wrong? Just enough to reproduce this and the error message being in the wrong location? It would be a great test case as we try to improve errors showing up in the right place.

**Comment by cyorobert** [comment]
Sure, I was specifying that the calling function returned `Promise<Array<Object>>` rather than `Promise<Array<any>>`.

Example:

```javascript
const fetchMyDatabaseModel = (primary_key : number) : Promise<Array<Object>> => {
    return mapBookshelfCollection(MyBookshelfModel.where({
        'id' : primary_key
    }).fetchAll())
}
```

**Comment by samwgoldman** [comment]
Is the error message any better on master (in try flow)? I'm not sure how to repro.

**Comment by cyorobert** [comment]
@samwgoldman I wasn't able to reproduce it in try flow either on 0.55. The issue I had can be boiled down to the following scenario: TL;DR, I believe this issue can probably be closed as it's not reproducible for me either in the sandbox environment.

I have function 1 and function 2. Function 1 returns the result of calling function 2 with an argument passed to it. Function 2 returns type A, but function 1 was incorrectly labeled to return type B when it should have been labeled to return type A. Flow correctly identified an error, but told me the error was with function 1 when the error should have been called on function 2.

Here is an example of the scenario I just tried to illustrate (working correctly on all available try flow versions): https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgCCYAvGAN6phgAWeM8YAXGAM4YBOAlgHYDmYVAF902fGABCpClTAI4HGABNmbTrwHD0qYMACSPMBhpdWYJXgC2cADRG6YPBw4KwpsAGMFHPB4wwsMBwYAEMPPBU6Hzkogi4MNho4AFdlMAAjPFQoZJ4-LjhDQoIACgBKVSlKajAfDGSOQwx5EoAiAAl6RigXSzAcvIwCop48VrKtHX1DS0CQnGCuDxChwrA9ORCzH3CuADcNMBDDJxcOMEs8VlYQvgJYuwQ6GcDOZIJT1wQtuXiafty+TWxWygNWTXkYBKl2utwILHY3H4FRYxGq1DqDUM5FoXTgqhhNzuYBEQiAA

Attempting to reproduce the original issue on try flow:
https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoVAXAngBwKZgBCccA1gM4AWeMUAsnACY1gC8YA3qmD2AIYYMAJwCWAIwCuGPOTAAufgDssqAL6ZcBYmSo0oAYTgwYeAMYYRcRW07deAWyY1ZCgIJChfLAB5tFarQMzDAAfGroeAAeOHBCGGCmVuTx9nw4frq0hsZmFlY2ABRgGNSKfGIm8mAACkJw9iLkeL4k-nrZJuaWiiFgAJRVtfWNze6ePnzKIb2svVy8YEJ4GBJC1iV4ZRV4AHQbigUFiTld+QoZAQZGnXmKA24eXi06l0E0M73HN907jsGyAH4Etdcj8-s4qgBtAC6fTsCwRiJ4e1Kh3BMBcYDGTwuejeoQGszA6PIvzSaKcMEJvXROwEwnEUhkfThSLZCx2pgEpkoh2pYBhcPUqGAwDoWDAjXIEgICD4smoSwANGB8htJfYcCZ7JsMAJusTlpQmDswABJMBy2TkfCmERQLAiRQAc2KlAExIlIk12t1+vySxWazwjBFwBqdQaTW82J8AHkxAArXLTS2lSXxQOrRSydVLaUweJwKDEtK4rIg07WBCUEQ8yWyIZR0aPCZTMKJHMpLBmn14HWKPW3OhGpiFe4R4bR2PeBPJ8yponzHid5JgRgSez2LAAfUYAj4O+SQgk5lWBHOrUyVxOtxsHBJUI49NEkmkmI4AXgVQARAhYjAjA-qoqjQqoYDwqu8Qbluu44JGIw2E2Iw7PmRgAG54AUMHbnuB5HsIp5BngrI8FmaylukV6XB0oJWNhm64fBU4kWoQA

**Comment by SamChou19815** [comment]
The issue was fixed a long time ago
